File: LeetCode/Python/solutions/p0382.py
import random
from typing import Optional

from Python.auxiliary.ListNode import ListNode


class Solution:

    def __init__(self, head: Optional[ListNode]):
        #
        # Solution v1: Brute Force
        #
        # Runtime: 340 ms @ (beats) 5.2%
        # Memory Usage: 17.1 MB @ (beats) 97.21%
        #
        self.vals = []

        while head:
            self.vals.append(head.val)
            head = head.next

        #
        # Solution v1.1: Performance improvement
        #
        #
        # Runtime: 69 ms @ (beats) 85.87%
        # Memory Usage: 17.6 MB @ (beats) 6.13%
        #
        # self.n = len(self.vals) - 1

    def getRandom(self) -> int:
        #
        # Solution v1: Brute Force
        #
        # rand = random.Random()
        # return self.vals[rand.randint(0, self.n)]

        #
        # Solution v1.1: Performance improvement
        #
        pos = int(random.random() * len(self.vals))
        return self.vals[pos]

File: LeetCode/Java/src/solutions/p0382_LinkedListRandomNode/P0382.java
package solutions.p0382_LinkedListRandomNode;

import auxiliary.ListNode;

import java.util.ArrayList;
import java.util.List;

public class P0382 {

    List nodes = new ArrayList<Integer>();

    public void Solution(ListNode head) {
        /*
         * Solution v1: Brute Force
         *
         * Runtime: 12 ms @ (beats) 48.76%
         * Memory Usage: 45 MB @ (beats) 14.89%
         */
        while (head != null) {
            nodes.add(head.val);
            head = head.next;
        }

    }

    public int getRandom() {
        int n = (int) (Math.random() * this.nodes.size());
        return (int) this.nodes.get(n);
    }
}
